### PR TITLE
Try to ensure usize marker does not get merged

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This follows up on [this conversation](https://github.com/rust-lang/rust/pull/69209#discussion_r379911282). However, I'm not confident this is quite correct, so feedback is appreciated, as always.

### CODE REVIEW

File: src/libcore/fmt/mod.rs [hunk 1]
```diff
@@ -265,7 +265,15 @@ pub struct ArgumentV1<'a> {
...
 #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
-static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |_, _| loop {};
+static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |ptr, _| {
+    // This tries to ensure we don't collapse into some other function; we can't
+    // guarantee that doesn't happen, but this tries to make sure that if we
+    // *do* get collapsed with some other function, that function also takes a
+    // `usize` as it's first argument, ensuring correctness below when we read
+    // out the associated value from `ArgumentV1`.
+    let _v: usize = *crate::hint::black_box(ptr);
+    loop {}
```
Comment (eddyb) [inline src/libcore/fmt/mod.rs]: If you make this return `Ok(())`, you could actually call it before reading the `usize` from `Argument`, ensuring that what this closure does is guaranteed to happen, giving more confidence that if there was UB, it was invoked by calling the `fn` pointer which should be safe to call.

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -265,7 +265,15 @@ pub struct ArgumentV1<'a> {
...
 // arguments), so this is merely an additional check.
 #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
-static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |_, _| loop {};
+static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |ptr, _| {
+    // This tries to ensure we don't collapse into some other function; we can't
+    // guarantee that doesn't happen, but this tries to make sure that if we
+    // *do* get collapsed with some other function, that function also takes a
+    // `usize` as it's first argument, ensuring correctness below when we read
+    // out the associated value from `ArgumentV1`.
+    let _v: usize = *crate::hint::black_box(ptr);
```
Comment (eddyb) [inline src/libcore/fmt/mod.rs]: My suggestion was to `black_box(*ptr)`. Or you could do a `volatile_load` if you want.

The goal is to make this more specific to "the reference is valid as an `&usize`", which requires the `load` to actually happen (meaning its result must be used).

File: src/libcore/fmt/mod.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/libcore/fmt/mod.rs]: It's a bit annoying to call the formatter as we don't have a `&mut Formatter<'_>` around and I'd rather not manufacture one just for this. I think we could likely thread it through but I think that hurts readability and such so I'd rather not.

**Comment by Mark-Simulacrum** [comment]
Changed to `black_box(*ptr)`.

File: src/libcore/fmt/mod.rs (hunk 1 above)
Comment (eddyb) [reply src/libcore/fmt/mod.rs]: Oh, right, that method doesn't take the formatter, fair enough.

File: src/libcore/fmt/mod.rs [hunk 3]
```diff
@@ -265,7 +265,15 @@ pub struct ArgumentV1<'a> {
...
 // arguments), so this is merely an additional check.
 #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
-static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |_, _| loop {};
+static USIZE_MARKER: fn(&usize, &mut Formatter<'_>) -> Result = |ptr, _| {
+    // This tries to ensure we don't collapse into some other function; we can't
+    // guarantee that doesn't happen, but this tries to make sure that if we
+    // *do* get collapsed with some other function, that function also takes a
+    // `usize` as it's first argument, ensuring correctness below when we read
+    // out the associated value from `ArgumentV1`.
+    let _v: usize = crate::hint::black_box(*ptr);
```
Comment (Centril) [inline src/libcore/fmt/mod.rs]: Are we using `black_box` here for functional correctness?

Comment (Mark-Simulacrum) [reply src/libcore/fmt/mod.rs]: Yes, we were, I think I agree we probably shouldn't be. Switched to a volatile read.

Comment (Centril) [reply src/libcore/fmt/mod.rs]: Thank you :)

**Review by eddyb — APPROVED** [review]

**Pull request merged**